### PR TITLE
Decouple train/val/test code by using separate pipelines for each. Remove --change_lr since --lr_scheduler can perform same behavior.

### PR DESCRIPTION
**Pull request by samiwilf**
Summary: Decouple train/val/test code by using separate pipelines for each. Remove --change_lr since --lr_scheduler can perform same behavior.  Cleanup comments and code.

Differential Revision: D41264865

### CODE REVIEW

File: torchrec_dlrm/dlrm_main.py [hunk 1]
```diff
@@ -499,75 +406,54 @@ class TrainValTestResults:
 
 def train_val_test(
     args: argparse.Namespace,
-    train_pipeline: TrainPipelineSparseDist,
+    pipeline: TrainPipelineSparseDist,
```
Comment (colin2328) [inline torchrec_dlrm/dlrm_main.py:409]: since we will be creating new pipelines inside here, IMO it makes more sense to create the main train_pipeline in this function as well (instead of passing in train_pipeline and creating more pipelines inside)

File: torchrec_dlrm/dlrm_main.py [hunk 2]
```diff
@@ -499,75 +406,54 @@ class TrainValTestResults:
...
         TrainValTestResults.
     """
 
     train_val_test_results = TrainValTestResults()
 
-    train_iterator = iter(train_dataloader)
-    test_iterator = iter(test_dataloader)
     for epoch in range(args.epochs):
-        val_iterator = iter(val_dataloader)
+        new_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
```
Comment (colin2328) [inline torchrec_dlrm/dlrm_main.py:433]: we use new_pipeline here, but never use pipeline passed in. why?

File: torchrec_dlrm/dlrm_main.py [hunk 3]
```diff
@@ -499,75 +406,54 @@ class TrainValTestResults:
...
-            test_iterator if epoch == args.epochs - 1 else train_iterator
-        )
-        val_accuracy, val_auroc = _evaluate(
-            args.limit_val_batches,
-            train_pipeline,
-            val_iterator,
-            val_next_iterator,
-            "val",
-        )
+        new_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
```
Comment (colin2328) [inline torchrec_dlrm/dlrm_main.py:447]: call this val_pipeline?

File: torchrec_dlrm/dlrm_main.py [hunk 4]
```diff
@@ -391,98 +351,45 @@ def _train(
...
+        pbar = tqdm(iter(int, 1), desc=f"Epoch {epoch}", total=len(train_dataloader), disable=False)
     for it in itertools.count():
         try:
-            if is_rank_zero:
-                if print_lr:
-                    for i, g in enumerate(train_pipeline._optimizer.param_groups):
-                        print(f"lr: {it} {i} {g['lr']:.6f}")
-            train_pipeline.progress(combined_iterator)
+            if validation_freq and it > 0 and it % validation_freq == 0:
+                new_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
```
Comment (colin2328) [inline torchrec_dlrm/dlrm_main.py:382]: here we will be running full val_pipeline over val_dataloader every validation_freq. Is this intended?

Comment (colin2328) [reply torchrec_dlrm/dlrm_main.py:382]: is it necessary to create so many pipelines? why not reuse same and just pass in new iters over the dataloaders?

File: torchrec_dlrm/dlrm_main.py (hunk 3 above)
Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py]: The reason for the variable name `new_pipeline` is to communicate to viewers that the pipelines created for train, test, and validation are the same.  If we had names val_pipeline, train_pipeline, test_pipeline, I could see that raising the question in people's minds what's the difference between a val_pipeline and a test_pipeline, for example.

File: torchrec_dlrm/dlrm_main.py [hunk 5]
```diff
@@ -391,98 +355,43 @@ def _train(
...
+        pbar = tqdm(iter(int, 1), desc=f"Epoch {epoch}", total=len(train_dataloader), disable=False)
     for it in itertools.count():
         try:
-            if is_rank_zero:
-                if print_lr:
-                    for i, g in enumerate(train_pipeline._optimizer.param_groups):
-                        print(f"lr: {it} {i} {g['lr']:.6f}")
-            train_pipeline.progress(combined_iterator)
+            if validation_freq and it > 0 and it % validation_freq == 0:
+                _evaluate(limit_val_batches, new_pipeline(pipeline), val_dataloader, "val")
```
Comment (colin2328) [inline torchrec_dlrm/dlrm_main.py:386]: do we need to create a new_pipeline each iteration? why not have one eval pipeline? (and just pass in new iter of the dataloader)?

File: torchrec_dlrm/dlrm_main.py (hunk 4 above)
Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py]: It's necessary to create pipelines in this way if we want to keep the code this clean.   Also I'm not seeing impact on performance when comparing the mlperf dcn 1-hot benchmark timings before and after this change.

Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py]: > here we will be running full val_pipeline over val_dataloader every validation_freq. Is this intended?

Yes this is intended.

File: torchrec_dlrm/dlrm_main.py (hunk 5 above)
Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py:386]: If we used the same pipeline for intra-epoch validation, it complicates ordering batches correctly.  When StopIteration is reached in _evaluate, `two_filler_batches` are in the pipeline.  The these two batches are the first two batches of the validation set. That means subsequent intra-epoch validation runs would need to slice off the first two batches from the validation iterator since the first two batches are already in the pipeline.  Additionally, args.limit_val_batches will need to be subtracted by 2 (the pipeline length - 1) when used to limit the iterator.    These differences in behavior between the first run and subsequent runs means there would need to be conditionals expressing if epoch > do this, otherwise do that.  The above changes would increase and complicate the code.  We aren't in a hurry, so we can think about it.  

One note. Under the hood, PyTorch uses already-allocated GPU memory for the intra-validation pipeline buffer because [`with torch.no_grad()` reduces memory consumption](https://pytorch.org/docs/stable/generated/torch.no_grad.html). The reduced memory consumption translates into available already-allocated GPU memory for buffered batches of newly spawned eval-only pipelines.

File: torchrec_dlrm/dlrm_main.py (hunk 2 above)
Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py]: The original idea was to treat `pipeline` as the source of the context object and have `pipeline` be in charge of setting the contexts of other pipelines created.

I have refactored the code more, and in the newer code I use `pipeline` as shown below. I refactored by creating the function:
```
def new_pipeline(pipeline: TrainPipelineSparseDist) -> TrainPipelineSparseDist:
    empty_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
    empty_pipeline._context = pipeline._context
    return empty_pipeline
```
and using this function such as:
```
    for epoch in range(args.epochs):
        _train(
            new_pipeline(pipeline),
            train_dataloader,
            val_dataloader,
            epoch,
            args.batch_size,
            lr_scheduler,
            args.print_lr,
            args.validation_freq_within_epoch,
            args.limit_train_batches,
            args.limit_val_batches,
        )
        val_accuracy, val_auroc = _evaluate(args.limit_val_batches, new_pipeline(pipeline), val_dataloader, "val")
```

R

File: torchrec_dlrm/dlrm_main.py (hunk 3 above)
Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py]: A reasonable counter view is that val_pipeline, train_pipeline, test_pipeline would communicate which pipeline pairs with which dataloader, but I do not see a risk of mistaking one pipeline for another that would raise the need to create three pipeline variable names.

File: torchrec_dlrm/dlrm_main.py (hunk 5 above)
Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py:386]: > do we need to create a new_pipeline each iteration? why not have one eval pipeline? (and just pass in new iter of the dataloader)?

I'll scratch my first paragraph above regarding reusing the same pipeline for validation.   It can be done cleanly if we add `pipeline._connected = False` at the start of `def _evaluate`.   This way, we could use the same pipeline for all non-training runs, which means validation and test runs.

Comment (samiwilf) [reply torchrec_dlrm/dlrm_main.py:386]: However, the underscore in _connected is the PEP-8 internal use indicator. That generally means not to tamper with such variables. @colin2328 what are you thinking?

**Comment by samiwilf** [comment]
This has been merged. 
See https://github.com/facebookresearch/dlrm/commit/37da857a406655fbf3218a94530f97deb6793769
